app/models/topic.rb

```diff
@@ -285,12 +285,22 @@ def recover!(recovered_by = nil)
     JOIN group_users gu ON gu.user_id = :user_id AND gu.group_id = tg.group_id
   SQL
 
-  scope :private_messages_for_user, ->(user) do
-    private_messages.where(
-      "topics.id IN (#{PRIVATE_MESSAGES_SQL_USER})
-      OR topics.id IN (#{PRIVATE_MESSAGES_SQL_GROUP})",
-      user_id: user.id
-    )
+  scope :private_messages_for_user, ->(user, opts = {}) do
+    private_message_scope = private_messages
+
+    sql = <<~SQL
+    topics.id IN (#{PRIVATE_MESSAGES_SQL_USER})
+    OR topics.id IN (#{PRIVATE_MESSAGES_SQL_GROUP})
+    SQL
+
+    if opts[:custom_sql_condition]
```

Comment:
This custom_sql_condition feels a little strange, and IMO increases the risk someone will introduce an SQL injection vulnerability in future. Why can't the consumer just chain another `.where` or `.or` onto the scope like

```
Topic.private_messages_for_user(user).or("my custom SQL")
```

Reply:
> Why can't the consumer just chain another .where or .or onto the scope like

We can't just do `or("my custom SQL")` as the method doesn't work like this. We have to pass an ActiveRecord relation to it which makes the code look something like this: 

```
    new_scope = scope.private_messages_for_user(user)

    if is_moderator?
      new_scope = new_scope.or(scope.where(<<~SQL))
        OR topics.subtype = '#{TopicSubtype.moderator_warning}'
        OR topics.id IN (#{Topic.has_flag_scope.select(:topic_id).to_sql})
      SQL
    end

    new_scope
```

Personally, I'm not a big fan of using ActiveRecord constructs when the query becomes more complex. In this case though, I didn't want to alter the existing `Topic.secured` and `Topic.private_messages_for` scopes so I've made the changes to alter the scope using AR constructs.